### PR TITLE
Feature/cart

### PR DESCRIPTION
**Pull request by Cristianjs93**
# Shopping Cart Feature

## Description

This pull request introduces the Shopping Cart feature to our application. Users can now add products to their cart directly from a restaurant's page by clicking the "Add" button. Within the cart, users can modify the quantity of products they want. We've also implemented control to prevent adding products from different restaurants when there are already active items in the cart. All the information is correctly saved in the data base usin an api service to create the order. Redux has been incorporated to manage states, and the SweetAlert library is used to provide user notifications at each step of the process.

## Changes Made

- Implemented the Shopping Cart feature.
- Added the "Add" button funcionality to allow users to add products to their cart.
- Enabled users to modify product quantities within the cart.
- Implemented logic to prevent adding products from different restaurants to the cart if there are already items present.
- Integrated Redux for state management.
- Incorporated SweetAlert for user notifications during the process.
- Order created in the data base.

## Checklist

- [x] Shopping Cart feature added.
- [x] "Add" button implemented for adding products.
- [x] Quantity modification functionality added.
- [x] Logic to prevent adding products from different restaurants.
- [x] Redux integrated for state management.
- [x] SweetAlert library for user notifications.
- [x] Api service to create order in db.

### Adding products
image

### Checkout
image

### Order created
image

### CODE REVIEW

File: src/components/Cart/index.jsx [hunk 1]
```diff
@@ -1,21 +1,110 @@
...
-          className="order-cart__empty__image"
-        />
-        <p className="order-cart__empty__paragraph">
-          Looks like you have not order something. Go ahead, order some yummy
-          food from menu..
-        </p>
-      </div>
+    <div className='order-cart'>
+      <h4 className='order-cart__empty__title'>Cart Items:</h4>
+      {products.length == 0 ? (
```
Comment (ht1204) [inline src/components/Cart/index.jsx]: `{!products.length`

File: src/components/Cart/index.jsx [hunk 2]
```diff
@@ -1,21 +1,110 @@
...
+                    id={product.id}
+                    className='order-cart__fill__box__button'
+                    onClick={handleIncrement}>
+                    <i
+                      id={product.id}
+                      className='bi bi-plus order-cart__fill__box__button__image'
+                    />
+                  </button>
+                </div>
+                {product.quantity == 0 && (
```
Comment (ht1204) [inline src/components/Cart/index.jsx]: `{!product.quantity`

File: src/components/OrderOnline/index.jsx [hunk 3]
```diff
@@ -32,14 +45,21 @@ const OrderOnline = ({ menu }) => {
...
                 <p className='restaurant-order__menu__item-description'>
                   {product.description}
                 </p>
                 <h5>${product.price}.00</h5>
-                <button className='restaurant-order__menu__item-button'>
+                <button
+                  id={restaurantId}
+                  className='restaurant-order__menu__item-button'
+                  disabled={
+                    products.find((item) => item.id == product.id)
```
Comment (ht1204) [inline src/components/OrderOnline/index.jsx]:
```js
 disabled={products.find((item) => item.id == product.id)}
```

File: src/main.jsx [hunk 4]
```diff
@@ -1,15 +1,19 @@
...
+import { RouterProvider } from 'react-router-dom';
+import router from './router/index.jsx';
+import { Provider } from 'react-redux';
+import store from './store/redux/index.js';
+import './index.scss';
+import mapboxgl from 'mapbox-gl';
 
 mapboxgl.accessToken =
-  "pk.eyJ1IjoiY3Jpc3RpYW5qczkzIiwiYSI6ImNsa292YXBvcjBhcDIzcmw5cThnOGl6MTMifQ.K0CoYwD-yvtbr2VDTFecXQ";
+  'pk.eyJ1IjoiY3Jpc3RpYW5qczkzIiwiYSI6ImNsa292YXBvcjBhcDIzcmw5cThnOGl6MTMifQ.K0CoYwD-yvtbr2VDTFecXQ';
```
Comment (ht1204) [inline src/main.jsx]: this access token should be importable through a constant

File: src/pages/SingleRestaurant/index.jsx [hunk 5]
```diff
@@ -66,7 +74,24 @@ const SingleRestaurant = () => {
         <div className='single-restaurant__principal__cart'>
           <div className='single-restaurant__principal__cart__sticky'>
             <div className='single-restaurant__principal__cart__items'>
-              <Cart />
+              <Cart>
+                <div className='order-cart__fill__bottom__buttons-box'>
+                  <button
+                    className='order-cart__fill__bottom__buttons-box__button'
+                    onClick={handleClear}>
+                    clear
```
Comment (ht1204) [inline src/pages/SingleRestaurant/index.jsx]: Clear

File: src/pages/SingleRestaurant/index.jsx [hunk 6]
```diff
@@ -66,7 +74,24 @@ const SingleRestaurant = () => {
...
+                  </button>
+                  <Link
+                    to='/checkout'
+                    className='order-cart__fill__bottom__buttons-box__link'>
+                    <button
+                      id={restaurant.id}
+                      className='order-cart__fill__bottom__buttons-box__button'>
+                      place order
```
Comment (ht1204) [inline src/pages/SingleRestaurant/index.jsx]: Place Order

File: src/pages/Summary/index.jsx [hunk 7]
```diff
@@ -19,31 +25,44 @@ const Checkout = () => {
...
+      <div className='summary__checkout-container'>
+        <div className='summary__contentca'>
+          <div className='summary__cartfirst'>
+            <div className='summary__cart-container'>
+              <Cart>
+                <div className='order-cart__fill__bottom__buttons-box'>
+                  <button
+                    className='order-cart__fill__bottom__buttons-box__button'
+                    onClick={handleClear}>
+                    clear
```
Comment (ht1204) [inline src/pages/Summary/index.jsx]: Clear

File: src/store/redux/slices/cartSlice.js [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        product.id == payload.id
+          ? {
+              ...product,
+              quantity: (product.quantity += payload.quantity),
+            }
+          : product;
+      });
+
+      state.products = state.products.filter(
+        (product) => product.quantity != 0
```
Comment (ht1204) [inline src/store/redux/slices/cartSlice.js]: `(product) => product.quantity !== 0`

File: src/store/redux/slices/cartSlice.js [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+      let total = 0;
+
+      state.products.forEach(
+        (product) => (total += product.price * product.quantity)
+      );
+
+      state.payment = total;
+
+      if (state.products.length == 0) {
```
Comment (ht1204) [inline src/store/redux/slices/cartSlice.js]: `if (!state.products.length)`

File: src/utils/toast/toast.scss [hunk 10]
```diff
@@ -0,0 +1,7 @@
+.custom-toast {
+  display: flex !important;
```
Comment (ht1204) [inline src/utils/toast/toast.scss:2]: guess there was not a better choice

Comment (Cristianjs93) [reply src/utils/toast/toast.scss:2]: No sir. sweetalert styles block another specified properties

File: src/components/Cart/index.jsx (hunk 1 above)
Comment (Cristianjs93) [reply src/components/Cart/index.jsx]: resolved

File: src/components/Cart/index.jsx (hunk 2 above)
Comment (Cristianjs93) [reply src/components/Cart/index.jsx]: resolved

File: src/main.jsx (hunk 4 above)
Comment (Cristianjs93) [reply src/main.jsx]: resolved

File: src/pages/SingleRestaurant/index.jsx (hunk 5 above)
Comment (Cristianjs93) [reply src/pages/SingleRestaurant/index.jsx]: resolved

File: src/pages/SingleRestaurant/index.jsx (hunk 6 above)
Comment (Cristianjs93) [reply src/pages/SingleRestaurant/index.jsx]: resolved

File: src/components/OrderOnline/index.jsx (hunk 3 above)
Comment (Cristianjs93) [reply src/components/OrderOnline/index.jsx]: resolved

File: src/pages/Summary/index.jsx (hunk 7 above)
Comment (Cristianjs93) [reply src/pages/Summary/index.jsx]: resolved

**Review by jesusdavid24 — APPROVED** [review]

**Pull request merged**